File: repository/src/main/java/hu/farago/repo/model/entity/mongo/EarningsCallFlat.java
package hu.farago.repo.model.entity.mongo;

import java.io.Serializable;
import java.math.BigInteger;

import org.joda.time.DateTime;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

import hu.farago.repo.model.entity.mongo.embedded.HTone;

/**
 * The original flattener script is:
 * 
 * 	db.earnings_call.aggregate( [
	   { $project : { 
	       _id: 1, 
	       tradingSymbol : 1, 
	       publishDate : 1, 
	       wordSize: { $size: { "$ifNull": [ "$words", [] ] }}, 
	       "h_tone.positiveCount": 1, 
	       "h_tone.negativeCount": 1,
	       q_and_a_wordSize: { $size: { "$ifNull": [ "$qAndAWords", [] ] }}, 
	       "q_and_a_h_tone.positiveCount": 1, 
	       "q_and_a_h_tone.negativeCount": 1,
	       }
	   },
	   { $out: "earnings_call_flat" }
	]);
 * 
 * @author neural
 *
 */


@Document(collection = "earnings_call_flat")
public class EarningsCallFlat implements Serializable {
	
	private static final long serialVersionUID = -1538987715086324171L;

	@Id
	public BigInteger id;
	public String tradingSymbol;
	public DateTime publishDate;
	public int wordSize;
	@Field("h_tone")
	public HTone hTone;
	public int q_and_a_wordSize;
	@Field("q_and_a_h_tone")
	public HTone qAndAHTone;
	
}

File: other/earningsCallFlat.js
db.earnings_call.aggregate( [
	{ $match: 
		{
			publishDate: { "$exists": true, "$ne": null }
		}
	},
	{ $project : 
		{ 
			_id: 1, 
			tradingSymbol : 1, 
			publishDate: 1,
			date_number : {$add :[{$multiply:[{ $year: "$publishDate" }, 10000]}, {$multiply:[{ $month: "$publishDate" }, 100]}, { $dayOfMonth: "$publishDate" }]},
			time_number : {$add :[{$multiply:[{ $hour: "$publishDate" }, 10000]}, {$multiply:[{ $minute: "$publishDate" }, 100]}, { $second: "$publishDate" }]},
			wordSize: { $size: { "$ifNull": [ "$words", [] ] }}, 
			"h_tone.positiveCount": { "$ifNull": [ "$h_tone.positiveCount", -1 ] }, 
			"h_tone.negativeCount": { "$ifNull": [ "$h_tone.negativeCount", -1 ] },
			q_and_a_wordSize: { $size: { "$ifNull": [ "$qAndAWords", [] ] }}, 
			"q_and_a_h_tone.positiveCount": { "$ifNull": [ "$q_and_a_h_tone.positiveCount", -1 ] }, 
			"q_and_a_h_tone.negativeCount": { "$ifNull": [ "$q_and_a_h_tone.negativeCount", -1 ] },
		}
	},
	{ $out: "earnings_call_flat" }
]);